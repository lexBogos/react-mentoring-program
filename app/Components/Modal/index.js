import React from 'react';
import './index.scss';

const Modal = ({active, setActive, fields, movieListHook, movieList}) => {
                    return <div className={active ? 'modal modalActive': 'modal'}>
                        <div className='modal__content' onClick={e => e.stopPropagation()}>
                            <a href='#' className='closeModal' onClick={(e) => {e.preventDefault(); setActive(false)}}>âœ–</a>
                            {fields.func(movieListHook, movieList, setActive, fields.index)}
                        </div>
                    </div>}

export default Modal;