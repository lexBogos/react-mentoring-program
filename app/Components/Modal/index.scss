.modal{
    height: 100vh;
    width: 100vw;
    background-color: rgba(0, 0, 0 , 0.4);
    position: fixed;
    top: 0;
    left: 0;
    display: flex;
    align-items: center;
    justify-content: center;
    opacity: 0;
    pointer-events: none;
    transition: 0.5s;
    z-index: 999999;
}

.modal__content{
    padding: 20px;
    border-radius: 5px;
    background-color: #232323;
    height: 90%;
    width: 30%;
}

.modalActive{
    opacity: 1;
    pointer-events: all;
}

.closeModal{
    color:rgb(255, 255, 255);
    text-decoration: none;
    position: relative;
    left: 97%;
    top:-2%;
    font-weight: 400;
}
.closeModal:visited {
    color:rgb(255, 255, 255);
    font-weight: 400;
}
.addMovie{
    display: flex;
    flex-direction: column;
}
.addMovie>div>span{
    color:#f65261;
    font-size: 20px;
    left:10%;
}

.addMovie>div{
    margin-left:10%;
    width: 80%;
    display: flex;
    flex-direction: column;

}

.addMovie>div>input{
    margin-top: 10px;
    margin-bottom: 20px;
    width: 100%;
    font-size: 20px;
}

.addMovie>div>select{
    margin-top: 10px;
    margin-bottom: 20px;
    width: 100%;
    font-size: 20px;
    width: 100%;
    height: 45px;
    border-radius: 3px;
    overflow: hidden;
    opacity: 0.7;
    background-color: #555555;
}

.addMovie>h3{
    margin-left: 10%;
    font-weight: 400;
}

.addMovie>div>button{
    border: none;
    border-radius: 3%;
    color: #fff;
    cursor: pointer;
    font-size:16px;
    font-weight: 400;
    line-height: 45px;
    margin: 0 0 2em;
    max-width: 160px; 
    position: relative;
    text-decoration: none;
    text-transform: uppercase;
    width: 100%; 
    background-color: rgb(246,82,97);
}

.addMovie>div>button:active {
    background-color: rgb(246,82,97);
    box-shadow: 0 5px #666;
    transform: translateY(4px);
  }

.modalButton{
    flex-direction: row !important;
    justify-content: space-between;
}
