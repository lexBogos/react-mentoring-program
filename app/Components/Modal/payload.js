import React from "react";
import { Formik, Form, Field } from "formik";
import * as Yup from 'yup';

const EditMovieModalFields = ({
  onMovieEdit,
  onModalDisable,
  data,
  movieStoreService,
}) => {

  const EditMovieSchema = Yup.object().shape({
    title: Yup.string()
      .min(2, 'Too Short!')
      .max(50, 'Too Long!')
      .required('Required'),
    description: Yup.string()
      .min(2, 'Too Short!')
      .max(200, 'Too Long!')
      .required('Required'),
    movieUrl: Yup.string().required('Required'),
    realiseDate: Yup.number()
      .required()
      .positive()
      .integer()
      .required('Required'),
    rate: Yup.number()
      .required()
      .positive()
      .required('Required'),
  });

  return (
    <>
      <h3>Edit Movie</h3>
      <Formik
        initialValues={{
          title: data.title,
          description: data.description,
          genre: data.genre,
          movieUrl: data.movieUrl,
          realiseDate: data.realiseDate,
          rate: data.rate,
        }}
        validationSchema={EditMovieSchema}
        onSubmit={values => {
          setTimeout(() => {
            movieStoreService.updateMovieData({movieId: data.id, movieData: values}).then((res)=>{
              console.log(res);
              onMovieEdit({...values, id: data.id});
              onModalDisable();
            });
          }, 0);
        }}
      >
        {({handleSubmit, values, errors}) => (
          <Form onSubmit={handleSubmit} className="addMovie">
            <Field
              name="title"
              value={values.title}
            />
            {errors.title && <div>{errors.title}</div>}
            <Field name="description" value={values.description} />
            {errors.description && <div>{errors.description}</div>}
            <Field name="movieUrl" value={values.movieUrl} />
            {errors.movieUrl && <div>{errors.movieUrl}</div>}
            <Field name="realiseDate" value={values.realiseDate} />
            {errors.realiseDate && <div>{errors.realiseDate}</div>}
            <Field
              as="select"
              name="genre"
              value={values.genre.split(",")[0]}
            >
              <option value="Action">Action</option>
              <option value="Biography">Biography</option>
              <option value="Crime">Crime</option>
              <option value="Drama">Drama</option>
              <option value="Sci-Fi">Sci-Fi</option>
            </Field>
            <Field name="rate" value={values.rate} />
            {errors.rate && <div>{errors.rate}</div>}
            <button type="submit">Submit</button>
            <button onClick={() => {onModalDisable()}}>RESET</button>
          </Form>
        )}
      </Formik>
    </>
  );
};

const DeleteMovieModalFields = ({
  onMovieRemove,
  chosenMovieId,
  movieStoreService,
}) => {
  return (
    <div className="addMovie">
      <h3>Delete Movie</h3>
      <div>
        <span>Are you sure want to delete this movie ?</span>
      </div>
      <div className="modalButton">
        <button
          onClick={() => {
            movieStoreService
              .removeMovie({ movieId: chosenMovieId })
              .then((res) => {
                console.log(res);
              });
            onMovieRemove(chosenMovieId);
          }}
        >
          CONFIRM
        </button>
      </div>
    </div>
  );
};

const AddMovieModalFields = ({ onMovieAdd, movieStoreService }) => {
  const movieObj = {
    id: Math.round(Math.random() * 1000000000) + "", //Replace with UUID
    title: "",
    realiseDate: "",
    movieUrl: "",
    genre: "sci-fi",
    overview: "",
    runTime: "",
  };

  const AddMovieSchema = Yup.object().shape({
    title: Yup.string()
      .min(2, 'Too Short!')
      .max(50, 'Too Long!')
      .required('Required'),
    description: Yup.string()
      .min(2, 'Too Short!')
      .max(200, 'Too Long!')
      .required('Required'),
    movieUrl: Yup.string().required('Required'),
    realiseDate: Yup.number()
      .required()
      .positive()
      .integer()
      .required('Required'),
    rate: Yup.number()
      .required()
      .positive()
      .required('Required'),
  });

  return (
    <>
      <h3>Edit Movie</h3>
      <Formik
        initialValues={{
          id: Math.round(Math.random() * 1000000000) + "", //Replace with UUID
          title: "",
          description: "",
          realiseDate: "",
          movieUrl: "",
          genre: "sci-fi",
          rate: "",
        }}
        validationSchema={AddMovieSchema}
        onSubmit={values => {
          setTimeout(() => {
            movieStoreService.updateMovieData({ movieData: values }).then((res)=>{
              console.log(res);
              onMovieAdd(values);
            });
          }, 0);
        }}
      >
        {({handleSubmit, values, errors}) => (
          <Form onSubmit={handleSubmit} className="addMovie">
            <Field
              name="title"
              value={values.title}
            />
            {errors.title && <div>{errors.title}</div>}
            <Field name="description" value={values.description} />
            {errors.description && <div>{errors.description}</div>}
            <Field name="movieUrl" value={values.movieUrl} />
            {errors.movieUrl && <div>{errors.movieUrl}</div>}
            <Field name="realiseDate" value={values.realiseDate} />
            {errors.realiseDate && <div>{errors.realiseDate}</div>}
            <Field
              as="select"
              name="genre"
              value={values.genre.split(",")[0]}
            >
              <option value="Action">Action</option>
              <option value="Biography">Biography</option>
              <option value="Crime">Crime</option>
              <option value="Drama">Drama</option>
              <option value="Sci-Fi">Sci-Fi</option>
            </Field>
            <Field name="rate" value={values.rate} />
            {errors.rate && <div>{errors.rate}</div>}
            <button type="submit">Submit</button>
          </Form>
        )}
      </Formik>
    </>
  );
};

export { AddMovieModalFields, EditMovieModalFields, DeleteMovieModalFields };
