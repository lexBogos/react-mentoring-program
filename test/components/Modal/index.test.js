import React from 'react';
import  Modal from '../../../app/Components/Modal/index';
import '@testing-library/jest-dom/extend-expect';
import { createStore } from "redux";
import reducer from "../../../app/reducers/index";
import { Provider } from "react-redux";
import { render as rtlRender } from '@testing-library/react';



function render(
  ui,
  {
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

describe('Modal component', () => {

    let closeModal;
    let preventDefault;

    beforeEach(()=>{
        closeModal = jest.fn();
        preventDefault = jest.fn();
    });

    const movies = [
        {
            title: 'Goodfellas',
            id: 'tt0099685',
            realiseDate: 1990,
            rate: 4.9,
            genre: 'Biography, Crime, Drama',

        },
        {
            title: 'Goodfellas',
            id: 'tt0133093',
            realiseDate: 1999,
            rate: 5
        },
        {
            title: 'Terminator 2: Judgment Day',
            id: 'tt0103064',
            realiseDate: 1991,
            rate: 5
        },
        {
            title: 'Modern Times',
            id: 'tt0027977',
            realiseDate: 1936,
            rate: 4.2
        },
        {
            title: 'Memento',
            id: 'tt0209144',
            realiseDate: 2000,
            rate: 4
        },
        {
            title: 'Dune',
            id: 'tt1160419',
            realiseDate: 2021,
            rate: 3.8
        }, 
      ];

    
    it('should return component without content if there is no modalActionType', () => {

        const { asFragment } = render(<Modal movies={movies} closeModal={closeModal} modalActive={true}/>, { initialState: { user: 'Redux User' } });

        expect(asFragment(<Modal/>)).toMatchSnapshot();
    });

    it('should return component with EditMovieModalFields component if modalActionType is "EDIT_MOVIE"', () => {

        const { asFragment } = render(<Modal/>, { initialState: { movies:movies, closeModal:closeModal, modalActive:true, chosenMovieIndex: 0, modalActionType: 'EDIT_MOVIE' } });

        expect(asFragment(<Modal/>)).toMatchSnapshot();
    });

    it('should return component with DeleteMovieModalFields component if modalActionType is "REMOVE_MOVIE"', () => {

        const { asFragment } = render(<Modal/>, { initialState: { closeModal:closeModal, modalActive:true, chosenMovieIndex: 0, modalActionType: 'REMOVE_MOVIE' } });

        expect(asFragment(<Modal/>)).toMatchSnapshot();
    });

    it('should return component with AddMovieModalFields component if modalActionType is "MOVIE_ADD"', () => {

        const { asFragment } = render(<Modal/>, { initialState: { closeModal:closeModal, modalActive:true, chosenMovieIndex: 0, modalActionType: 'MOVIE_ADD' } });

        expect(asFragment(<Modal/>)).toMatchSnapshot();
    });
})