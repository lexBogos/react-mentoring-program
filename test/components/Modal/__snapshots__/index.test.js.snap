// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should return component with AddMovieModalFields component if modalActionType is "MOVIE_ADD" 1`] = `
<DocumentFragment>
  <div
    class="modal modalActive"
  >
    <div
      class="modal__content"
    >
      <a
        class="closeModal"
        href="#"
      >
        ✖
      </a>
      <h3>
        Add Movie
      </h3>
      <form
        action="#"
        class="addMovie"
      >
        <input
          name="title"
          value=""
        />
        <input
          name="description"
          value=""
        />
        <input
          name="movieUrl"
          value=""
        />
        <input
          name="realiseDate"
          value=""
        />
        <select
          name="genre"
        >
          <option
            value="Action"
          >
            Action
          </option>
          <option
            value="Biography"
          >
            Biography
          </option>
          <option
            value="Crime"
          >
            Crime
          </option>
          <option
            value="Drama"
          >
            Drama
          </option>
          <option
            value="Sci-Fi"
          >
            Sci-Fi
          </option>
        </select>
        <input
          name="rate"
          value=""
        />
        <button
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Modal component should return component with DeleteMovieModalFields component if modalActionType is "REMOVE_MOVIE" 1`] = `
<DocumentFragment>
  <div
    class="modal modalActive"
  >
    <div
      class="modal__content"
    >
      <a
        class="closeModal"
        href="#"
      >
        ✖
      </a>
      <div
        class="addMovie"
      >
        <h3>
          Delete Movie
        </h3>
        <div>
          <span>
            Are you sure want to delete this movie ?
          </span>
        </div>
        <div
          class="modalButton"
        >
          <button>
            CONFIRM
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Modal component should return component with EditMovieModalFields component if modalActionType is "EDIT_MOVIE" 1`] = `
<DocumentFragment>
  <div
    class="modal modalActive"
  >
    <div
      class="modal__content"
    >
      <a
        class="closeModal"
        href="#"
      >
        ✖
      </a>
      <h3>
        Edit Movie
      </h3>
      <form
        action="#"
        class="addMovie"
      >
        <input
          name="title"
          value="Goodfellas"
        />
        <input
          name="description"
          value=""
        />
        <input
          name="movieUrl"
          value=""
        />
        <input
          name="realiseDate"
          value="1990"
        />
        <select
          name="genre"
        >
          <option
            value="Action"
          >
            Action
          </option>
          <option
            value="Biography"
          >
            Biography
          </option>
          <option
            value="Crime"
          >
            Crime
          </option>
          <option
            value="Drama"
          >
            Drama
          </option>
          <option
            value="Sci-Fi"
          >
            Sci-Fi
          </option>
        </select>
        <input
          name="rate"
          value="4.9"
        />
        <button
          type="submit"
        >
          Submit
        </button>
        <button
          type="button"
        >
          RESET
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Modal component should return component without content if there is no modalActionType 1`] = `
<DocumentFragment>
  <div
    class="modal"
  >
    <div
      class="modal__content"
    >
      <a
        class="closeModal"
        href="#"
      >
        ✖
      </a>
    </div>
  </div>
</DocumentFragment>
`;
